plugin/src/Reconciler/diff.spec.lua

```diff
@@ -0,0 +1,265 @@
+	]]
+	itFIXME("should ignore unreadable properties", function()
+		local knownInstances = InstanceMap.new()
+		local virtualInstances = {
+			ROOT = {
+				ClassName = "Fire",
+				Name = "Fire",
+				Properties = {
+					-- heat_xml is a serialization-only property that is not
+					-- exposed to Lua.
+					heat_xml = {
+						Type = "Float32",
+						Value = 5,
+					},
+				},
+				Children = {},
+			},
+		}
+
+		local rootInstance = Instance.new("Fire")
+		knownInstances:insert("ROOT", rootInstance)
+
+		local ok, patch = diff(knownInstances, virtualInstances, "ROOT")
+
+		Log.warn("{:#?}", patch)
+
+		assert(ok, tostring(patch))
+
+		assert(isEmpty(patch.removed))
+		assert(isEmpty(patch.added))
+		assert(isEmpty(patch.updated))
+	end)
+
+	it("should generate a patch removing unknown children by default", function()
+		local knownInstances = InstanceMap.new()
+		local virtualInstances = {
+			ROOT = {
+				ClassName = "Folder",
+				Name = "Folder",
+				Properties = {},
+				Children = {},
+			},
+		}
+
+		local rootInstance = Instance.new("Folder")
+		local unknownChild = Instance.new("Folder")
+		unknownChild.Parent = rootInstance
+		knownInstances:insert("ROOT", rootInstance)
+
+		local ok, patch = diff(knownInstances, virtualInstances, "ROOT")
+
+		assert(ok, tostring(patch))
+
+		assert(isEmpty(patch.added))
+		assert(isEmpty(patch.updated))
+		expect(#patch.removed).to.equal(1)
+		expect(patch.removed[1]).to.equal(unknownChild)
+	end)
+
+	it("should generate an empty patch if unknown children should be ignored", function()
+		local knownInstances = InstanceMap.new()
+		local virtualInstances = {
+			ROOT = {
+				ClassName = "Folder",
+				Name = "Folder",
+				Properties = {},
+				Children = {},
+				Metadata = {
+					ignoreUnknownInstances = true,
+				},
+			},
+		}
+
+		local rootInstance = Instance.new("Folder")
+		local unknownChild = Instance.new("Folder")
+		unknownChild.Parent = rootInstance
+		knownInstances:insert("ROOT", rootInstance)
+
+		local ok, patch = diff(knownInstances, virtualInstances, "ROOT")
+
+		assert(ok, tostring(patch))
+
+		assert(isEmpty(patch.added))
+		assert(isEmpty(patch.updated))
+		assert(isEmpty(patch.removed))
+	end)
+
+	it("should generate a patch with an added child", function()
+		local knownInstances = InstanceMap.new()
+		local virtualInstances = {
+			ROOT = {
+				ClassName = "Folder",
+				Name = "Folder",
+				Properties = {},
+				Children = {"CHILD"},
+			},
+
+			CHILD = {
+				ClassName = "Folder",
+				Name = "Child",
+				Properties = {},
+				Children = {},
+			},
+		}
+
+		local rootInstance = Instance.new("Folder")
+		knownInstances:insert("ROOT", rootInstance)
+
+		local ok, patch = diff(knownInstances, virtualInstances, "ROOT")
+
+		assert(ok, tostring(patch))
+
+		assert(isEmpty(patch.updated))
+		assert(isEmpty(patch.removed))
+		expect(size(patch.added)).to.equal(1)
+		expect(patch.added["CHILD"]).to.equal(virtualInstances["CHILD"])
+	end)
+end
```

Comment:
An alternative test case where you have a cyclic reference may also be good. I doubt this can come up during normal use however it would be nice to ensure it doesn't blow up rojo in the event it does somehow happen.

Reply:
I'm not too worried about this case coming up. I think we can avoid adding the extra machinery to detect cycles unless it's something that could feasibly happen.